README.rst
.. contents ::

WARNING
=======

As of version 3.0, zopeskel is no longer a package of its own. All the
templates that make it work have been moved into packages in the templer
namespace. Some of the templates you may be expecting are no longer available.
If you require older templates, please be sure to install ZopeSkel<3.0

Introduction
============

ZopeSkel provides a collection of project templates for Plone and Zope
development projects.

ZopeSkel uses the `paster <http://pythonpaste.org/script/>`_ Python library
internally.

Installing ZopeSkel
===================

ZopeSkel can be installed in one of two ways: with `buildout
<http://www.buildout.org/>`_ or with `virtualenv <http://virtualenv.org/>`_.

.. note ::

    Despite existing documentation to the contrary, it is not recommended to
    install ZopeSkel in your system python.

Buildout installation
---------------------------

Here are instructions how to include ZopeSkel as a part of your
local ``buildout.cfg``.

* You can use zopeskel to add new projects to your buildout ``src/`` folder.

Add to your ``buildout.cfg``::

    parts =
       ...
       zopeskel

    [zopeskel]
    # installs paster and Zopeskel
    recipe = zc.recipe.egg
    eggs =
       PasteScript
       ZopeSkel

After re-running buildout, you will have ``zopeskel`` and ``paster``
commands in the ``bin`` directory of your buildout.

Virtualenv installation
-----------------------

First, install virtualenv into your system::

    easy_install virtualenv

Next, create a virtual environment with the new ``virtualenv`` command::

    virtualenv --no-site-packages --distribute zopeskelenv

Once virtualenv is finished, you can install zopeskel to your new virtual
environment::

    zopeskelenv/bin/easy_install zopeskel

Once this is complete, you will be left with ``zopeskel`` and ``paster``
commands in the ``bin`` directory inside your virtualenv.

Available Templates
===================

To see details of the available templates::

    zopeskel --list

More info about how zopeskel works::

    zopeskel --help

Using Templates
===============

Creating a package using virtualenv ZopeSkel installation::

    source zopeskelenv/bin/activate
    zopeskel <template_name> <package_name>

The folder created (``<package_name>``) can be checked in to the versioning
system of your choice.

Local Commands
==============

Local commands are available for the ``plone_basic`` and ``archetype``
package templates.  If you create a package using one of these templates you
will be provided with information about the presence of local commands.  

To run a local command and add further features to your package, you will
run the paster command ``add`` from within your package.  Your current working
directory must be at or inside the directory where you find your package's 
egg-info, so you can do the following::

    $ bin/templer plone_basic my.package
    ....
    $ cd my.package/src
    $ ../../bin/paster --help
    ...
    Templer local commands:
      add  Allows the addition of further templates to an existing package
    $ ../../bin/paster add --list
    browserlayer:  A Plone browserlayer
    browserview:   A browser view skeleton
    $ ../../bin/paster add browserlayer
    ...

Developing ZopeSkel
===================

If you wish to contribute to the zopeskel project we welcome your
contribution. Zopeskel is now distributed with its own built-in buildout, so
to begin, all you need to do is check out the source, bootstrap with your
desired version of python, and run bin/buildout.

Since all of the template that are provided by ZopeSkel are now in templer
namespace packages, the ZopeSkel package uses mr.developer to provide access
to all the templer packages required in the src directory of the buildout.
Development should take place in those packages. There should be no templates
in the ZopeSkel package at all.

To get started, simply clone the zopeskel repository to your local machine,
bootstrap with your preferred python, and run the buildout::

    git clone git@github.com:collective/ZopeSkel.git zopeskel
    cd zopeskel
    python2.6 bootstrap.py
    ...
    bin/buildout

Testing
-------

Since version 1.5, ZopeSkel has tests. It's required to run these before you
check in any changes you make. You should run the full test suite in both
Python 2.4 and Python 2.6, as both versions are in common use among Zope and
Plone developers. They can be run like so::

    bin/test

Please ensure that all tests pass in Python 2.4 and Python 2.6 before making
any checkins to any templer package used by zopeskel.

Fixing Bugs
-----------

There are a number of open issues in the queue at
http://plone.org/products/zopeskel/issues and your help is always welcome in
closing any you feel competent to take on. Please note that there is a
zopeskel mailing list, so if you have any questions about your approach to
fixing a bug, you should post to the list first.

More info
=========

Issue tracker

* http://plone.org/products/zopeskel/issues

Source code

* https://github.com/collective/zopeskel

Mailing List

* https://lists.plone.org/mailman/listinfo/plone-zopeskel

Please contribute by submitting patches for what you consider 'best of
breed' file layouts for starting Zope and Plone projects.


HISTORY.txt
Changelog
=========

3.0b3 - May 22, 2012
--------------------

- fix bad package release
  [cewing]

3.0b2 - May 22, 2012
--------------------

- fix package documentation error
  [cewing]

3.0b1 - May 22, 2012
--------------------

- restore localcommands for ``plone_basic`` and ``archetypes`` templates
  [cewing]

- add ``plone_nested`` template, providing a simple nested-namespace package
  suitable for Plone development
  [cewing]

3.0a1 - October 7, 2011
-----------------------

- depend on templer packages, remove all local templates, update README
  [cewing]

2.21.2 - October 6, 2011
------------------------

* fixing bad bad release fixer [cewing]

2.21.1 - October 6, 2011
------------------------

* fixing bad release [cewing]

2.21 - October 6, 2011
----------------------

* Formatting changes, pep8 fixes, add buildout template warning to
  plone4_buildout [cewing]

* Updated default Plone 3 version to 3.3.6 [cewing]

* Fixed some wording and made the instructions for using zopeskel in buildout
  line up with the unified installer method.  We should reduce the number of
  'conflicting' ways we have displayed in the wild.  [cewing]

* Fixed plone4_buildout of trying to checkout collective.docs from the old 
  location [miohtama]

* Bumped Plone version to 4.1 [miohtama]

2.20 - August 9, 2011
---------------------

* Remove buildout.threatlevel extension from plone4_buildout. [davisagli]

* Made plone4_buildout simpler: now it only asks for Plone version [miohtama]

* Various references to Internet documentation [miohtama]

* Created a recommended path how ZopeSkel templates should be used to create 
  a new Plone site project [miohtama]

* Bumped Plone 4 version to 4.0.5 [miohtama]

* Edit plone3_buildout help text, fixes:
  http://plone.org/products/zopeskel/issues/59
  [aclark]

2.19 - January 21, 2011
--------------------------

* In every setup.py, replace '%3A' with ':', making list_classifier url an url
  which can actually be used.
  [thet]

* Added plone4_buildout [miohtama]

* Improvements to README [miohtama]

2.18 - September 30, 2010
-------------------------

* Add bin/zopeskel --version flag.  Reports current version of zopeskel
  (http://plone.org/products/zopeskel/issues/48)
  [cewing]

* plone3_buildout: use Plone 3.3.5 by default.
  [maurits]

* Fixed a problem with the generic setup profile in plone templates in which a
  profile was registered in zcml, but no profiles directory was created.
  (http://plone.org/products/zopeskel/issues/46)
  [cewing]

* use i18n:domain plone for portlet title so it need not be translated twice
  (http://plone.org/products/zopeskel/issues/44)
  [fRiSi]

* use the correct name of package, so it can be used in package with 2 dots or
  1 dots in package name
  [lucmult]

2.17 - June 14, 2010
--------------------

* Added a pin for PasteScript>=1.7.2.  This solves an issue where using zopeskel
  with an early version of PasteScript resulted in a method signature mismatch
  on the paster command 'challenge' method.  Fixes issue #42
  (http://plone.org/products/zopeskel/issues/42)
  [cewing]

* Fixed lingering problem in Archetypes template with registering a 'locales'
  directory before one is created.  Users now need to add an i18nlocale via
  addcontent in order to get this directory and have it registered via zcml.
  [cewing]

* Fixed dependency in Archetypes template doctests on the presence of the log-in
  portlet.  Tests now open the 'login_form' instead.  Resolves issue #40
  (http://plone.org/products/zopeskel/issues/40)
  [cewing]

* Added new browserlayer localcommand for plone and archetype templates.
  [marcosfromero]

* Added stub locales directory for plone3_theme template, resolving issue #33
  (http://plone.org/products/zopeskel/issues/33)
  [cewing]

* Set the GenericSetup profile version number in the generated
  metadata.xml to 1000, which is better for the alphabetical ordering
  that GS does.
  [maurits]

* Added "environment-vars = zope_i18n_compile_mo_files true" to
  plone3_buildout template. This is enabled only for Plone 4 (eggifiedzope is
  true).
  [vincentfretin]

2.16 - March 22, 2010
---------------------

* Fixed syntax error in generated portlet constructor code
  [ajung]

* Added a new 'BoundedIntVar' variable type.  It validates that the provided
  argument is both an integer and between provided min and max values (inclusive
  on both ends).
  [cewing]

* Updated archetype and plone3_portlet templates to comply with pep8 and
  pyflakes validation at all stages of a skeleton build process.  Resolves the
  issue here: (http://plone.org/products/zopeskel/issues/24)
  [cewing]

* Added explanatory text to the plone_hosting template regarding the fact that
  it is unsuitable for use with plone versions later than 3.1.7, fixed
  http://plone.org/products/zopeskel/issues/25
  [cewing]

* Fixed the problem with generated archetypes content type meta_type names
  documented here:
  (http://plone.org/products/zopeskel/issues/22)
  [cewing]

* Updated tests for the plone template to include tests for the fixes to message
  factory, zcml sub-package include and i18n translation registration problems
  below
  [cewing]

* Cleaned up a number of issues with defining and importing a package-wide
  MessageFactory in plone template local commands:
  (http://plone.org/products/zopeskel/issues/29)
  [cewing]

* Cleaned up a number of issues with importing sub-package zcml files after
  using local commands in the plone template:
  (http://plone.org/products/zopeskel/issues/28)
  [cewing]

* Fixed an issue with the plone template zcmlmeta local command to ensure that
  the proper directive was inserted into the top-level configure.zcml
  [cewing]

* Fixed the bug in the plone template that caused zcml errors when the
  i18nlocale local command had not been run:
  (http://plone.org/products/zopeskel/issues/26)
  [cewing]

* Improved i18n for portlets
  (http://plone.org/products/zopeskel/issues/31)
  [fRiSi]


zopeskel/hosting.py
import errno
import os
import socket
import subprocess
import sys
import copy

from paste.script import templates
from zopeskel.base import BadCommand
from zopeskel.base import BaseTemplate, EASY, EXPERT
from zopeskel.vars import var, IntVar, BooleanVar, StringVar
from zopeskel import abstract_buildout

plone25s = {
        "2.5.5": "https://launchpad.net/plone/2.5/2.5.5/+download/Plone-2.5.5.tar.gz",
        "2.5.4": "https://launchpad.net/plone/2.5/2.5.4/+download/Plone-2.5.4-2.tar.gz",
        "2.5.3": "https://launchpad.net/plone/2.5/2.5.3/+download/Plone-2.5.3-final.tar.gz",
        "2.5.2": "http://plone.googlecode.com/files/Plone-2.5.2-1.tar.gz",
        "2.5.1": "http://heanet.dl.sourceforge.net/sourceforge/plone/Plone-2.5.1-final.tar.gz",
        "2.5"  : "http://heanet.dl.sourceforge.net/sourceforge/plone/Plone-2.5.tar.gz",
        }

class StandardHosting(abstract_buildout.AbstractBuildout):
    _template_dir = "templates/plone_hosting"
    use_cheetah = True
    summary = "Plone hosting: buildout with ZEO and Plone versions below 3.2"
    required_templates = []
    help = """
This template helps you to create an entire zope hosting setup, including
ZEO and a single Zope client instance.  If you desire, it can also install
and set up the Varnish Caching/Proxy Server.    

Please Note:
Due to changes in the packaging of Plone, this template is not suitable
for versions of Plone beyond 3.1.7.  If you are trying to use a later 
version of plone, you will need to edit the buildout resulting from this
template in order to have it work correctly.  Information related to these
issues may be found here:

http://plone.org/documentation/kb/repair-a-plone-3.1-buildout

and here:

http://plone.org/products/zopeskel/issues/25
"""

    vars = copy.deepcopy(abstract_buildout.AbstractBuildout.vars);
    vars = [

        abstract_buildout.VAR_ZOPE_USER,
        abstract_buildout.VAR_ZOPE_PASSWD, 

        IntVar(
            "base_port", 
            title="Base Port #",
            description="# to use as base for Zope/ZEO/proxy ports",
            modes=(EASY, EXPERT),
            page="Main",
            default=8000,
            help="""
For standardization, rather than selecting ports for Zope, ZEO, and
a proxy individually, these are tied together numerically.

ZEO port = Base + 0 | Proxy port = Base + 1 | HTTP port = Base + 10

If the ports specified by any of these numbers are already in use or
otherwise unavailable, this template will inform you of the problem and
exit with an error.  If this happens, please try another number for
'Base Port #'
""",
            ),
            
        BooleanVar(
            "proxy",
            title="Install proxy server?",
            description="Should a proxy server be installed?",
            default=False,
            help=""" 
If you ask for a proxy server to be installed, this template will include
the Varnish Caching/Proxy server.  If you wish to use a different proxy
server, please answer False and install your own.  
""",
            ),

        StringVar(
            "plone",
            title="Plone Version",
            description="Version to install (2.5, 2.5.1, 3.0, 3.0.1, etc)",
            default="3.1.7",
            help="""
You can use this template to install any version of Plone from 2.5 on. 
Versions of Plone more recent than 3.1.7 will require some editing of the
generated configuration files.  Please see the long description of this
template (run 'zopskel --list') for more details.  In general it is 
advisable to use the most recent version of Plone.  You can find a list of 
stable Plone releases at 

http://plone.org/products/plone/releases/

""",
            ),

        BooleanVar(
            "buildout",
            title="Run Buildout?",
            description="Should bin/buildout command be executed?",
            default=True,
            help="""
Would you like this template to automatically run the buildut command as soon
as it finishes creating the required files?  Please note that if you've chosen
to build a version of Plone more recent than 3.1.7 running buildout will fail
unless you make changes to the generated configuration files.  Please see the
long description of this package (run 'zopeskel --list') for more details.

If you intend on adding any specific third-party products or modifying the
buildout in any way, you should answer 'False'.  Then make your modifications
and run `python bootstrap.py` followed by `bin/buildout`.
""",
            ),

            ]

    def _buildout(self, output_dir):
        olddir=os.getcwd()
        try:
            os.chdir(output_dir)
            print "Bootstrapping the buildout"
            subprocess.call([sys.executable, "bootstrap.py"])
            print "Configuring the buildout"
            subprocess.call(["bin/buildout", "-n"])
        finally:
            os.chdir(olddir)

    def _checkPortAvailable(self, port):
        s=socket.socket()
        try:
            s.connect(("127.0.0.1", port))
        except socket.error, e:
            s.close()

            if e.args[0]==errno.ECONNREFUSED:
                return 

            raise BadCommand("Error checking port availability: %s" % e.args[1])

        s.close()
        raise BadCommand("Port %s is already in use" % port)


    def check_vars(self, vars, cmd):
        result=super(StandardHosting, self).check_vars(vars, cmd)

        base_port=result["base_port"]
        result["zeo_port"]=base_port
        result["proxy_port"]=base_port+1
        result["http_port"]=base_port+10

        self._checkPortAvailable(result["zeo_port"])
        self._checkPortAvailable(result["http_port"])
        if result["proxy"]:
            self._checkPortAvailable(result["proxy_port"])

        if vars["plone"] not in plone25s and not vars["plone"].startswith("3."):
            raise BadCommand("Unknown plone version: %s" % vars["plone"])

        return result

    def pre(self, command, output_dir, vars):
        vars["output_dir"]=os.path.abspath(output_dir)
        plone=vars["plone"]
        if plone.startswith("3."):
            vars["plone_recipe"]="plone.recipe.plone"
            vars["plone_recipe_version"]=plone
        else:
            vars["plone_recipe"]="plone.recipe.plone25install"
            vars["plone_url"]=plone25s[plone]
        super(StandardHosting, self).pre(command, output_dir, vars)

    def show_summary(self, vars):
        print
        print "Finished creation of standard hosting buildout."
        print
        print "Configuration summary:"
        print "  Plone     : %s" % vars["plone"]
        print
        print "  HTTP port : %s" % vars["http_port"]
        print "  ZEO port  : %s" % vars["zeo_port"]
        if vars["proxy"]:
            print "  Proxy port: %s" % vars["proxy_port"]
        else:
            print "  Proxy port: disabled"
        print
        print "  Zope admin user    :  %s" % vars["zope_user"]
        print "  Zope admin password:  %s" % vars["zope_password"]


    def post(self, command, output_dir, vars):
        output_dir=vars["output_dir"]
        if vars["buildout"]:
            self._buildout(output_dir)
        if not vars.get("hide_summary", False):
            self.show_summary(vars)
        super(StandardHosting, self).post(command, output_dir, vars)
